File: julia/Hand.jl
include("Card.jl")

mutable struct Hand
    cards::Vector{Card}  # List of cards in the hand
    unsorted::Vector{Card}  # Unsorted list of cards
    hand_type::String  # Type of hand (e.g., "pair", "three of a kind", etc.)
    key_cards::Vector{Card}  # Key cards in the hand
    kickers::Vector{Card}  # Kicker cards

    # Constructor for Hand struct
    function Hand(cards::Vector{Card})
        # Initialize the Hand with unsorted, cards, empty hand type, and empty key cards/kickers
        hand = new(cards, cards, "", Card[], Card[])
        
        return hand  # Return the constructed hand
    end
end

# Function to initialize the hand after creation
function initialize_hand!(hand::Hand)
    # Call your other functions here
    separate_key_cards_and_kickers!(hand)  # Separate key cards and kickers
    sort_hand!(hand)  # Sort the hand by card values
    evaluate_hand!(hand)  # Evaluate the hand type
end

function set_unsorted!(hand::Hand, cardList::Vector{Card})
    # Copy the cardList into the unsorted field
    hand.unsorted = copy(cardList)
end

# Function to generate a string representation of the hand
function toString(hand::Hand)
    for card in hand.unsorted
        if card!=nothing
            print(toString(card))
        end
    end
end

# Additional Helper functions
# Function to check if the hand is a flush
function is_flush(hand::Hand)
    suit = hand.cards[1].suit
    for card in hand.cards
        if card.suit != suit
            return false
        end
    end
    return true
end

# Function to check if the hand is a straight
function is_straight(hand::Hand)
    values = sort([Int(faceValue(card)) for card in hand.cards])  # Sort the face values

    # Check for high Ace straight (10-J-Q-K-A)
    if values[1] == 2 && values[5] == 14 && all(values .== [2, 3, 4, 5, 14])
        return true
    end

    # Check for regular straight
    for i in 2:5
        if values[i] != values[i-1] + 1
            return false
        end
    end
    return true
end

# Function to evaluate the hand type and set the hand values of each card
function evaluate_hand!(hand::Hand)
    face_counts = zeros(Int, 15)  # Array to count occurrences of each face value (index 0 is unused)
    
    for card in hand.cards
        face_counts[Int(faceValue(card))] += 1
    end

    flush = is_flush(hand)
    straight = is_straight(hand)

    if flush && straight
        if face_counts[14] > 0 && face_counts[13] > 0  # Check for Ace and King (Royal Flush)
            hand.hand_type = "Royal Flush"
            set_hand_value_to_cards!(hand, Float64(9000))
        else
            hand.hand_type = "Straight Flush"
            set_hand_value_to_cards!(hand, Float64(8000))
        end
    elseif contains_n_of_a_kind(face_counts, 4)
        hand.hand_type = "Four of a Kind"
        set_hand_value_to_cards!(hand, Float64(7000))
    elseif contains_n_of_a_kind(face_counts, 3) && contains_n_of_a_kind(face_counts, 2)
        hand.hand_type = "Full House"
        set_hand_value_to_cards!(hand, Float64(6000))
    elseif flush
        hand.hand_type = "Flush"
        set_hand_value_to_cards!(hand, Float64(5000))
    elseif straight
        hand.hand_type = "Straight"
        set_hand_value_to_cards!(hand, Float64(4000))
    elseif contains_n_of_a_kind(face_counts, 3)
        hand.hand_type = "Three of a Kind"
        set_hand_value_to_cards!(hand, Float64(3000))
    elseif count_pairs(face_counts) == 2
        hand.hand_type = "Two Pair"
        set_hand_value_to_cards!(hand, Float64(2000))
    elseif contains_n_of_a_kind(face_counts, 2)
        hand.hand_type = "One Pair"
        set_hand_value_to_cards!(hand, Float64(1000))
    else
        hand.hand_type = "High Card"
        set_hand_value_to_cards!(hand, Float64(0))
    end

    # After determining the hand type, separate into key cards and kickers
    separate_key_cards_and_kickers!(hand)
end

# Function to set hand values to cards (based on the hand type)
function set_hand_value_to_cards!(hand::Hand, base_value::Float64)
    for card in hand.cards
        hand_value = base_value + faceValue(card) + suitValue(card)
        # For "One Pair" and "Two Pair", don't add suitValue
        if hand.hand_type in ["One Pair", "Two Pair"]
            hand_value = base_value + faceValue(card)
        end
        setHandValue!(card, Float64(hand_value))
    end
end

# Custom comparison function for Hand objects
function Base.isless(hand1::Hand, hand2::Hand)
    # Use compare_hands to compare the two hands
    return compare_hands!(hand1, hand2) == -1  # If hand1 < hand2, return true
end

# Function to compare two hands based on the card hand_values and suits for tie-breaking
function compare_hands!(hand1::Hand, hand2::Hand)
    # First, compare the key cards by their hand values
    for i in 1:length(hand1.key_cards)
        card_comparison = compare_card_values(hand1.cards[i], hand2.cards[i])
        if card_comparison != 0
            return card_comparison
        end
    end
    
    for i in 1:length(hand1.cards)
        setHandValue!(hand1.cards[i] , suitValue(hand1.cards[i]));
    end
    
    for i in 1:length(hand2.cards)
        setHandValue!(hand2.cards[i] , suitValue(hand2.cards[i]));
    end

    # If key cards are equal, compare kicker cards
    for i in length(hand1.key_cards)+1:length(hand1.cards)
        card_comparison = compare_card_values(hand1.cards[i], hand2.cards[i])
        if card_comparison != 0
            return card_comparison
        end
    end
    
    return 0  # If all cards are equal, the hands are considered equal
end

# Helper function to compare card values (hand_value + suit value for tie-breaking)
function compare_card_values(card1::Card, card2::Card)
    card_value1 = card1.handValue
    card_value2 = card2.handValue

    if card_value1 != card_value2
        return card_value1 > card_value2 ? 1 : -1
    end

    # If the hand_value is the same, compare suits
    return compare_suits(card1, card2)
end

# Function to compare suits for tie-breaking
function compare_suits(card1::Card, card2::Card)
    suit_value1 = suitValue(card1)
    suit_value2 = suitValue(card2)

    if suit_value1 != suit_value2
        return suit_value1 > suit_value2 ? 1 : -1
    end

    return 0  # If suits are the same, cards are considered equal
end

# Function to sort the hand (key cards first, then kickers)
function sort_hand!(hand::Hand)
    # Sort key cards and kickers by face value and suit value
    sort!(hand.key_cards, by = card -> faceValue(card) + suitValue(card), rev=true)
    sort!(hand.kickers, by = card -> faceValue(card) + suitValue(card), rev=true)

    # Combine the sorted cards
    hand.cards = vcat(hand.key_cards, hand.kickers)

    # Check for Ace-to-5 straight ("wheel") and adjust if needed
    if is_straight(hand) && faceValue(hand.cards[1]) == 14 && faceValue(hand.cards[end]) == 2
        # Move the Ace to the end of the list to form the low straight
        hand.cards = vcat(hand.cards[2:end], hand.cards[1])
    end
end

# Function to check if there are n of a kind in the face counts
function contains_n_of_a_kind(face_counts, n)
    return any(count -> count == n, face_counts)
end

# Function to count pairs in the hand
function count_pairs(face_counts)
    return count(count -> count == 2, face_counts)
end

# Function to separate key cards and kickers based on frequency of face values
function separate_key_cards_and_kickers!(hand::Hand)
    face_counts = zeros(Int, 15)
    for card in hand.cards
        face_counts[Int(faceValue(card))] += 1
    end

    key_cards = Card[]
    kickers = Card[]

    for card in hand.cards
        if face_counts[Int(faceValue(card))] > 1
            push!(key_cards, card)
        else
            push!(kickers, card)
        end
    end

    hand.key_cards = key_cards
    hand.kickers = kickers
end

File: Go/Hand.go
package main

import (
	"fmt"
	"sort"
)

// Hand represents a poker hand, containing 5 cards
type Hand struct {
	cards       []*Card       // List of cards in the hand
	unsorted    []*Card       // Original unsorted list of cards
	handType    string        // Type of the hand (e.g., "Pair", "Full House")
	keyCards    []*Card       // Key cards that define the hand (e.g., the pair in a "One Pair")
	kickers     []*Card       // Remaining cards in the hand that don't contribute to the hand type
}

// NewHand creates a new hand from a list of cards
func NewHand(cards []*Card) *Hand {
	hand := &Hand{
		cards:    append([]*Card{}, cards...),
		unsorted: append([]*Card{}, cards...),
	}
	hand.separateKeyCardsAndKickers()
	hand.sortHand()
	hand.evaluateHand()
	return hand
}

// Create a custom type for a slice of *Hand pointers
type HandSlice []*Hand

// Implement sort.Interface for HandSlice
func (h HandSlice) Len() int {
	return len(h)
}

func (h HandSlice) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h HandSlice) Less(i, j int) bool {
	// Sort hands by rank (ascending order)
	// Change the comparison to "<" to reverse the order (descending)
	return h[i].CompareTo(h[j]) < 0
}

// GetCards returns the cards in the hand
func (h *Hand) GetCards() []*Card {
	return h.cards
}

// GetUnsortedCards returns the unsorted cards in the hand
func (h *Hand) GetUnsortedCards() []*Card {
	return h.unsorted
}

// SetUnsorted sets the unsorted list of cards for the hand
func (h *Hand) SetUnsorted(cards []*Card) {
	h.unsorted = append([]*Card{}, cards...)
}

// GetHandType returns the type of the hand (e.g., "Pair", "Straight")
func (h *Hand) GetHandType() string {
	return h.handType
}

// ToString returns a string representation of the hand
func (h *Hand) ToString() string {
	var result string
	for _, card := range h.unsorted {
		result += fmt.Sprintf("%3s ", card.String()) // Format each card in the hand
	}
	return result
}

// EvaluateHand determines the type of the hand (e.g., "Flush", "Straight", etc.)
func (h *Hand) evaluateHand() {
	// Initialize an array to count the occurrences of each face value
	faceCounts := make([]int, 15)
	for _, card := range h.cards {
		faceCounts[int(card.FaceValueNumeric())]++
	}

	// Check if the hand is a flush and/or a straight
	isFlush := h.isFlush()
	isStraight := h.isStraight()

	// Determine hand type and assign base value to each card
	if isFlush && isStraight {
		if faceCounts[14] > 0 && faceCounts[13] > 0 {
			h.handType = "Royal Flush"
			h.setHandValueToCards(1000)
		} else {
			h.handType = "Straight Flush"
			h.setHandValueToCards(900)
		}
	} else if h.containsNOfAKind(faceCounts, 4) {
		h.handType = "Four of a Kind"
		h.setHandValueToCards(800)
	} else if h.containsNOfAKind(faceCounts, 3) && h.containsNOfAKind(faceCounts, 2) {
		h.handType = "Full House"
		h.setHandValueToCards(700)
	} else if isFlush {
		h.handType = "Flush"
		h.setHandValueToCards(600)
	} else if isStraight {
		h.handType = "Straight"
		h.setHandValueToCards(500)
	} else if h.containsNOfAKind(faceCounts, 3) {
		h.handType = "Three of a Kind"
		h.setHandValueToCards(400)
	} else if h.countPairs(faceCounts) == 2 {
		h.handType = "Two Pair"
		h.setHandValueToCards(300)
	} else if h.containsNOfAKind(faceCounts, 2) {
		h.handType = "One Pair"
		h.setHandValueToCards(200)
	} else {
		h.handType = "High Card"
		h.setHandValueToCards(100)
	}
}

// SetHandValueToCards assigns a value to each card based on the hand type
func (h *Hand) setHandValueToCards(baseValue int) {
	for _, card := range h.cards {
		var handValue float64
		if h.handType != "One Pair" && h.handType != "Two Pair" {
			handValue = float64(baseValue) + card.FaceValueNumeric() + card.SuitValueNumeric()
		} else {
			handValue = float64(baseValue) + card.FaceValueNumeric()
		}
		card.SetHandValue(handValue)
	}
}

// IsFlush checks if the hand is a flush (all cards have the same suit)
func (h *Hand) isFlush() bool {
	suit := h.cards[0].Suit
	for _, card := range h.cards {
		if card.Suit != suit {
			return false
		}
	}
	return true
}

// IsStraight checks if the hand is a straight (cards form a sequence)
func (h *Hand) isStraight() bool {
	var values []int
	for _, card := range h.cards {
		values = append(values, int(card.FaceValueNumeric()))
	}
	sort.Ints(values)

	// Check for Ace-low straight (A-2-3-4-5)
	if(len(values)>=5){
    if values[0] == 2 && values[4] == 14 && values[1] == 3 && values[2] == 4 && values[3] == 5 {
  		return true
  	}
 }

	// Check for regular straight
 
	for i := 1; i < len(values); i++ {
		if values[i] != values[i-1]+1 {
			return false
		}
	}
	return true
}

// ContainsNOfAKind checks if the hand contains a certain number of cards with the same face value
func (h *Hand) containsNOfAKind(faceCounts []int, n int) bool {
	for _, count := range faceCounts {
		if count == n {
			return true
		}
	}
	return false
}

// CountPairs counts how many pairs are in the hand
func (h *Hand) countPairs(faceCounts []int) int {
	pairs := 0
	for _, count := range faceCounts {
		if count == 2 {
			pairs++
		}
	}
	return pairs
}

// SeparateKeyCardsAndKickers separates cards into key cards and kickers
func (h *Hand) separateKeyCardsAndKickers() {
	h.keyCards = []*Card{}
	h.kickers = []*Card{}

	faceCounts := make([]int, 15)
	for _, card := range h.cards {
		faceCounts[int(card.FaceValueNumeric())]++
	}

	for _, card := range h.cards {
		if faceCounts[int(card.FaceValueNumeric())] > 1 {
			h.keyCards = append(h.keyCards, card)
		} else {
			h.kickers = append(h.kickers, card)
		}
	}
}

// SortHand sorts the hand based on hand values and priority of key cards
func (h *Hand) sortHand() {
	// Sort key cards and kickers in descending order of hand value
	sort.Slice(h.keyCards, func(i, j int) bool {
		return h.keyCards[i].HandValue > h.keyCards[j].HandValue
	})
	sort.Slice(h.kickers, func(i, j int) bool {
		return h.kickers[i].HandValue > h.kickers[j].HandValue
	})

	// Combine the sorted key cards and kickers into the final hand
	h.cards = append(h.keyCards, h.kickers...)

	// Handle Ace-low straight (Ace used as low card in A-2-3-4-5)
	if h.isStraight() && h.cards[0].FaceValueNumeric() == 14 && h.cards[4].FaceValueNumeric() == 2 {
		h.cards = append(h.cards[1:], h.cards[0])
	}
}

// CompareTo compares this hand with another hand
func (h *Hand) CompareTo(other *Hand) int {
	for i := 0; i < len(h.keyCards); i++ {
		cardComparison := h.cards[i].HandValue - other.cards[i].HandValue
		if cardComparison != 0 {
			if cardComparison > 0 {
				return 1
			}
			return -1
		}
	}
 
 for i := range h.cards {
		h.cards[i].SetHandValue(h.cards[i].SuitValueNumeric()) // Modify HandValue using suitValue
	}
 
	// Compare remaining cards (kickers)
	for i := len(h.keyCards); i < len(h.cards); i++ {
		cardComparison := h.cards[i].HandValue - other.cards[i].HandValue
		if cardComparison != 0 {
			if cardComparison > 0 {
				return 1
			}
			return -1
		}
	}
	return 0 // Hands are equal if all comparisons are equal
}

